api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+  // The deployment config can be extended to provide platform specific configs.
+  // The supported config is [PipelineDeploymentConifg]().
+  google.protobuf.Any deployment_config = 3;
+
+  // The version of the sdk, which compiles the spec.
+  string sdk_version = 4;
+
+  // The version of the schema.
+  string schema_version = 5;
+
+  // The definition of the runtime parameter.
+  message RuntimeParameter {
+    // Required field. The type of the runtime parameter.
+    PrimitiveType.PrimitiveTypeEnum type = 1;
+    // Optional field. Default value of the runtime parameter. If not set and
+    // the runtime parameter value is not provided during runtime, an error will
+    // be raised.
+    Value default_value = 2;
+  }
+  // The runtime parameters of the pipeline. Keyed by parameter name.
+  map<string, RuntimeParameter> runtime_parameters = 6;
+}
+
+// The spec of task inputs.
+message TaskInputsSpec {
+  // The specification of a task input artifact.
+  message InputArtifactSpec {
+    // The name of the upstream task which produces the output that matches with
+    // the `output_artifact_key`.
+    string producer_task = 1;
+
+    // The key of [TaskOutputsSpec.artifacts][] map of the producer task.
+    string output_artifact_key = 2;
+  }
+
+  // Represents an input parameter. The value can be taken from an upstream
+  // task's output parameter (if specifying `producer_task` and
+  // `output_parameter_key`, or it can be a runtime value, which can either be
+  // determined at compile-time, or from a pipeline parameter.
+  message InputParameterSpec {
+    // Represents an upstream task's output parameter.
+    message TaskOutputParameterSpec {
+      // The name of the upstream task which produces the output parameter that
+      // matches with the `output_parameter_key`.
+      string producer_task = 1;
+
+      // The key of [TaskOutputsSpec.parameters][] map of the producer task.
+      string output_parameter_key = 2;
+    }
+    oneof kind {
+      // Output parameter from an upstream task.
+      TaskOutputParameterSpec task_output_parameter = 1;
+      // A constant value or runtime parameter.
+      ValueOrRuntimeParameter runtime_value = 2;
+    }
+  }
+
+  // A map of input parameters which are small values, stored by the system and
+  // can be queriable.
+  map<string, InputParameterSpec> parameters = 1;
+  // A map of input artifacts.
+  map<string, InputArtifactSpec> artifacts = 2;
+}
+
+// The spec of task outputs.
+message TaskOutputsSpec {
+  // The specification of a task output artifact.
+  message OutputArtifactSpec {
+    // The type of the artifact.
+    ArtifactTypeSchema artifact_type = 1;
+
+    // The properties of the artifact, which are determined either at
+    // compile-time, or at pipeline submission time through runtime parameters
+    map<string, ValueOrRuntimeParameter> properties = 2;
+
+    // The custom properties of the artifact, which are determined either at

```

Comment:
What are the use cases for this feature?

Can all the output artifact properties be part of the output type?

In KFP we declare outputs with complex types the following way:

```yaml
outputs:
- name: Output1
  type: {TensorflowSavedModel: {version: 2.2.1}}
```

Reply:
You can express the same data with artifact_type to be TensorflowSavedModel and a version string property or custom property.

Reply:
I'm trying to understand the difference and guidance regarding properties vs custom_properties.

I understand the reason why custom_properties exist in MLMD - they are used by frameworks to add system-related data that's not added by the user. But I wonder how does pipeline author decide which one to use.

Reply:
> I understand the reason why custom_properties exist in MLMD - they are used by frameworks to add system-related data that's not added by the user. But I wonder how does pipeline author decide which one to use.

Custom properties are used to set schemaless properties. I don't think it must be system related data. I will leave this question to Ajay. @neuromage